sparse/src/KokkosSparse_coo2crs.hpp

```diff
@@ -0,0 +1,344 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#include "KokkosSparse_CooMatrix.hpp"
+#include "KokkosSparse_CrsMatrix.hpp"
+#include "KokkosKernels_Utils.hpp"
+#include "Kokkos_UnorderedMap.hpp"
+#include <Kokkos_StdAlgorithms.hpp>
+
+#ifndef _KOKKOSSPARSE_COO2CRS_HPP
+#define _KOKKOSSPARSE_COO2CRS_HPP
+namespace KokkosSparse {
+namespace Impl {
+template <class DimType, class RowViewType, class ColViewType,
+          class DataViewType, bool InsertMode = false>
+class Coo2Crs {
+ private:
+  using RowViewScalarType  = typename RowViewType::value_type;
+  using ColViewScalarType  = typename ColViewType::value_type;
+  using DataViewScalarType = typename DataViewType::value_type;
+  using CrsST              = DataViewScalarType;
+  using CrsOT              = RowViewScalarType;
+  using CrsET              = typename DataViewType::execution_space;
+  using CrsMT              = void;
+  using CrsSzT             = ColViewScalarType;
+  using CrsType            = CrsMatrix<CrsST, CrsOT, CrsET, CrsMT, CrsSzT>;
+  using CrsValsViewType    = typename CrsType::values_type;
+  using CrsRowMapViewType  = typename CrsType::row_map_type::non_const_type;
+  using CrsColIdViewType   = typename CrsType::index_type;
+
+  using UmapValueViewType = Kokkos::View<CrsST *, CrsET>;
+  using UmapOpTypes =
+      Kokkos::UnorderedMapInsertOpTypes<UmapValueViewType, CrsOT>;
+  using UmapOpType = typename UmapOpTypes::AtomicAdd;
+
+  // Make public for Kokkos::View
+ public:
+  using UmapHasherType  = typename Kokkos::pod_hash<CrsOT>;
+  using UmapEqualToType = typename Kokkos::pod_equal_to<CrsOT>;
+  using UmapType = Kokkos::UnorderedMap<CrsOT, CrsST, CrsET, UmapHasherType,
+                                        UmapEqualToType>;
+  using UmapMemorySpace = typename UmapType::device_type::memory_space;
+
+  // Public for kokkos policies
+  struct coo2crsRp1 {};
+  struct rowmapRp1 {};
+  struct copyTp1 {};
+  struct copyRp1 {};
+
+  using copyTp1Pt         = Kokkos::TeamPolicy<copyTp1, CrsET>;
+  using copyTp1MemberType = typename copyTp1Pt::member_type;
+
+ private:
+  using BmapViewType = Kokkos::View<bool *, CrsET>;
```

Comment:
Yes, I think this would help in the near-term with issuing larger memory transactions. Going forward, we may want to use more of the encapsulated behavior of `Kokkos::BitSet`.